Improvments for topology figure script

chaehni opened the pull request:
This PR adds improvements to the script generating the Graphviz topology figure:
- Location labels can be added to ISDs/ASes via a flag
- The script now supports multiple links between the same ASes
- Links are annotated with their port information (via flag)
- Links and their corresponding Interface/port numbers are colored to make them distinguishable (via flag)
- The services running in an AS and their corresponding addresses are printed more nicely, making the graph easier to read

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/netsec-ethz/scion-web/126)
<!-- Reviewable:end -->

jonghoonkwon commented:
@chaehni, finally I had a chance to review your PR. I made some minor comments (see below), but overall, it looks good to me. I also tested the code, and it works for me. Great job!

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[utility/graphviz/graphviz_lib.py, line 108 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/126#-L81ygIVUohKSFib4Chk:-L81ygIVUohKSFib4Chl:bpusdaw) ([raw file](https://github.com/netsec-ethz/scion-web/blob/fe7084df52dca82afd62dcbdd4d2ccd9849e71da/utility/graphviz/graphviz_lib.py#L108)):*
> ```Python
>                 intra_isd_neighbor = \
>                     {'n_as': n_as, 'br-ip': border_routers[br]["Interfaces"][ie]["Public"]["Addr"], \
>                         'br-port': border_routers[br]["Interfaces"][ie]["Public"]["L4Port"],
> ```

the variable name `ie`gives me some confusion here. If the ie stands for the interface ID, lets use the name `if_id`

---

*[utility/graphviz/graphviz_lib.py, line 289 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/126#-L81zPos44NYynnhmzcW:-L81zPos44NYynnhmzcX:b2sxcl0) ([raw file](https://github.com/netsec-ethz/scion-web/blob/fe7084df52dca82afd62dcbdd4d2ccd9849e71da/utility/graphviz/graphviz_lib.py#L289)):*
> ```Python
>                     return (interface, port)
>         return ('','')
>     
> ```

Lets rid off the white spaces.

---

*[utility/graphviz/graphviz_lib.py, line 335 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/126#-L81zgWFuzQJXhlPM9MC:-L81zgWFuzQJXhlPM9MD:b-t5tat9) ([raw file](https://github.com/netsec-ethz/scion-web/blob/fe7084df52dca82afd62dcbdd4d2ccd9849e71da/utility/graphviz/graphviz_lib.py#L335)):*
> ```Python
>             for neighbor in self.AS["intra_n"]:
>                 n_dict = self.AS["intra_n"][neighbor]
>                 #br_id = str(n_dict["br-id"])
> ```

Lets rid off the comment line.

---

*[utility/graphviz/graphviz_lib.py, line 350 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/126#-L81zo7D-0KWu_RloWC2:-L81zo7D-0KWu_RloWC3:brhumzt) ([raw file](https://github.com/netsec-ethz/scion-web/blob/fe7084df52dca82afd62dcbdd4d2ccd9849e71da/utility/graphviz/graphviz_lib.py#L350)):*
> ```Python
>             and writes the info to the info and reverse info dict.
>             """
>         
> ```

Same here.

---

*[utility/graphviz/graphviz_lib.py, line 352 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/126#-L81zq-j5QTbI8ppXE7l:-L81zq-j5QTbI8ppXE7m:brhumzt) ([raw file](https://github.com/netsec-ethz/scion-web/blob/fe7084df52dca82afd62dcbdd4d2ccd9849e71da/utility/graphviz/graphviz_lib.py#L352)):*
> ```Python
>         
>             for interface in self.AS["inter_n"]:
>                 #n_dict = self.AS["inter_n"][interface]
> ```

Same here.

---

*[utility/graphviz/parse_folder.py, line 104 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/126#-L81x9c61AzxwZww8IUC:-L81x9c61AzxwZww8IUD:b7hoqwg) ([raw file](https://github.com/netsec-ethz/scion-web/blob/fe7084df52dca82afd62dcbdd4d2ccd9849e71da/utility/graphviz/parse_folder.py#L104)):*
> ```Python
> 
> def parse_desc_labels(gen_folder):
>     with open(gen_folder + '/../labels.json', 'r') as f:
> ```

I think we need an error handling code here for the case the label file is missing.

---

*[utility/graphviz/scion_topology_graphviz.py, line 52 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/126#-L82-7rmA4nsyPTSZoRm:-L82-7rmA4nsyPTSZoRn:b-5h08c7) ([raw file](https://github.com/netsec-ethz/scion-web/blob/fe7084df52dca82afd62dcbdd4d2ccd9849e71da/utility/graphviz/scion_topology_graphviz.py#L52)):*
> ```Python
>     c_neighbors = core_AS_list
>     n_neighbors = graph_lib.draw_edges_from_current(c_neighbors, isd_graph, ip_addresses)
>     #n_neighbors = graph_lib.draw_edges_from_current(non_core_AS_list, isd_graph, edge_labels)
> ```

same here.

---

*[utility/graphviz/scion_topology_graphviz.py, line 175 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/126#-L81wTtIj_MX5mnKeTeT:-L81wTtIj_MX5mnKeTeU:b-zi86xf) ([raw file](https://github.com/netsec-ethz/scion-web/blob/fe7084df52dca82afd62dcbdd4d2ccd9849e71da/utility/graphviz/scion_topology_graphviz.py#L175)):*
> ```Python
>     args = parser.parse_args()
>     topo = parse_gen_folder(args.gen_folder_path)
>     labels = parse_desc_labels(args.gen_folder_path)
> ```

It would be nice if we can specify the location of the label file rather than putting the file on the project root folder. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-web/126#-:-L81xX91z-A76riIb2Ry:bidw9p1)*
<!-- Sent from Reviewable.io -->

jonghoonkwon commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-web/126#-:-L82-Iy9TVq3prF16WXj:bnfp4nl)*
<!-- Sent from Reviewable.io -->

chaehni commented:
@jonghoonkwon Thanks for having a look. I addressed your comments.

---

Review status: 0 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[utility/graphviz/graphviz_lib.py, line 108 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/126#-L81ygIVUohKSFib4Chk:-L8CQfip7BXKvcStvIEC:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/fe7084df52dca82afd62dcbdd4d2ccd9849e71da/utility/graphviz/graphviz_lib.py#L108)):*
<details><summary><i>Previously, jonghoonkwon wrote…</i></summary><blockquote>

the variable name `ie`gives me some confusion here. If the ie stands for the interface ID, lets use the name `if_id`
</blockquote></details>

Done.

---

*[utility/graphviz/graphviz_lib.py, line 289 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/126#-L81zPos44NYynnhmzcW:-L8CQgHOzfTwK9z1umBL:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/fe7084df52dca82afd62dcbdd4d2ccd9849e71da/utility/graphviz/graphviz_lib.py#L289)):*
<details><summary><i>Previously, jonghoonkwon wrote…</i></summary><blockquote>

Lets rid off the white spaces.
</blockquote></details>

Done.

---

*[utility/graphviz/graphviz_lib.py, line 335 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/126#-L81zgWFuzQJXhlPM9MC:-L8CQvcOReoFsx7dYaOQ:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/fe7084df52dca82afd62dcbdd4d2ccd9849e71da/utility/graphviz/graphviz_lib.py#L335)):*
<details><summary><i>Previously, jonghoonkwon wrote…</i></summary><blockquote>

Lets rid off the comment line.
</blockquote></details>

Done.

---

*[utility/graphviz/graphviz_lib.py, line 350 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/126#-L81zo7D-0KWu_RloWC2:-L8CR3gGIGgIu8gY-EY2:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/fe7084df52dca82afd62dcbdd4d2ccd9849e71da/utility/graphviz/graphviz_lib.py#L350)):*
<details><summary><i>Previously, jonghoonkwon wrote…</i></summary><blockquote>

Same here.
</blockquote></details>

Done.

---

*[utility/graphviz/graphviz_lib.py, line 352 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/126#-L81zq-j5QTbI8ppXE7l:-L8CR9NGovzMma11_JqE:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/fe7084df52dca82afd62dcbdd4d2ccd9849e71da/utility/graphviz/graphviz_lib.py#L352)):*
<details><summary><i>Previously, jonghoonkwon wrote…</i></summary><blockquote>

Same here.
</blockquote></details>

Done.

---

*[utility/graphviz/parse_folder.py, line 104 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/126#-L81x9c61AzxwZww8IUC:-L8ESFd01F9b7t65a-5I:b-te7wk4) ([raw file](https://github.com/netsec-ethz/scion-web/blob/fe7084df52dca82afd62dcbdd4d2ccd9849e71da/utility/graphviz/parse_folder.py#L104)):*
<details><summary><i>Previously, jonghoonkwon wrote…</i></summary><blockquote>

I think we need an error handling code here for the case the label file is missing.
</blockquote></details>

I now added error handling for these cases:
- the file is missing
- the file is in the wrong format
- the file is incomplete

Additionally, I handled the case where the gen folder is missing.

---

*[utility/graphviz/scion_topology_graphviz.py, line 52 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/126#-L82-7rmA4nsyPTSZoRm:-L8CRVDkIWEYcH365Pke:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/fe7084df52dca82afd62dcbdd4d2ccd9849e71da/utility/graphviz/scion_topology_graphviz.py#L52)):*
<details><summary><i>Previously, jonghoonkwon wrote…</i></summary><blockquote>

same here.
</blockquote></details>

Done.

---

*[utility/graphviz/scion_topology_graphviz.py, line 175 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/126#-L81wTtIj_MX5mnKeTeT:-L8ESy7vD2wBE7ng6kYp:b-8c5a06) ([raw file](https://github.com/netsec-ethz/scion-web/blob/fe7084df52dca82afd62dcbdd4d2ccd9849e71da/utility/graphviz/scion_topology_graphviz.py#L175)):*
<details><summary><i>Previously, jonghoonkwon wrote…</i></summary><blockquote>

It would be nice if we can specify the location of the label file rather than putting the file on the project root folder. 
</blockquote></details>

I added a command line flag for specifying the location of the label file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-web/126#-:-L8ERzwiLgQ7OdN3J2RH:b-r8nrek)*
<!-- Sent from Reviewable.io -->
